Amazon S3 as an artifact deployment source

jonathanq opened the pull request:
I put together support to pull deployment artifacts from S3 in addition to Nexus, local files and URLs.  This is based entirely on the url provided - S3 artifacts must be in the form: `s3://bucket-name/path/to/object`

Included in this is a change to the format of the data_bags - I made it backwards compatible to current and previous formats, and wrote rspec tests around it as well.

I am fairly new to Ruby (and never done rspec tests before) - so suggestions or comments are welcome!

jonathanq committed:
Added S3 as a deployment artifact source using credentials in data_bags, ENV variables or IAM Instance Roles on EC2

jonathanq committed:
Fixed bucket and object existence checks and related rspec tests, stopped windows from trying to install libxml/libxslt

jonathanq committed:
Added another S3 reference in README.md

KAllan357 commented on providers/deploy.rb:
```diff
@@ -57,6 +57,21 @@ def load_current_resource
+  elsif from_s3?(@new_resource.artifact_location)
+    unless Chef::Artifact.windows?
+      %W{libxml2 libxslt libxml2-devel libxslt-devel}.each do |nokogiri_requirement|
+        package nokogiri_requirement do
```
Do you plan on using this logic with CentOS? I was driven crazy Nokogiri + CentOS so many times with unreproduceable failures. Some sort of timing error between installing these packages and then installing the C-extensions in the nokogiri gem. I realize there isn't much that can be done about this, since its a dependency of `aws-sdk`.

jonathanq replied:
I have been using a CentOS 6.4 vm locally for testing and didn't hit any issues.  I believe the reason for this (and the solution to your previous timing issues) is to do with the:

```
package package_name do
   action :install
end.run_action(:install)
```

The build-essential cookbook has a good description of this https://github.com/opscode-cookbooks/build-essential#usage

However looks like I combined both.  The `action :install` on the package is not needed since the `end.run_action(:install)` will install the packages at compile_time.  `chef_gem` is also a compile-time action.  I'll correct the `action :install` part since it should be `action :nothing` and re-test locally and amend this PR.

KAllan357 replied:
Sounds good. I knew the weird problems were related to the anatomy of the Chef run, and I could never reproduce them in a VM - it actually only ever happened to me on real boxes. Hopefully, this doesnt pop up again!

jonathanq replied:
I was able to reproduce the timing issue consistently on a CentOS VM by changing it to just:

```
package nokogiri_requirement do
   action :install
end
```

When I went back to the `end.run_action(:install)` method and ran it against a CentOS 6 EC2 instance I hit an issue with the vanilla image not having gcc and make installed.

Should we stick with just adding those 2 onto the list like: 

```
%W{gcc make libxml2 libxslt libxml2-devel libxslt-devel}.each do |nokogiri_requirement|
    package nokogiri_requirement do
      action :nothing
    end.run_action(:install)
  end
```

Or would it make sense to look at including the build-essential cookbook in this case?  I am not crazy about adding another cookbook dependency - but I also hate to re-invent the wheel either.

KAllan357 replied:
Yeah, :-1: on adding more dependencies. I suppose all those packages should be fine.

KAllan357 commented on libraries/chef_artifact.rb:
```diff
@@ -111,11 +125,55 @@ def get_actual_version(node, artifact_location, ssl_verify=true)
+          require 'aws-sdk'
+          config = config_for(node, DATA_BAG_AWS)
+          protocol, bucket_name, object_name = URI.split(source_file).compact
+          object_name = object_name[1..-1]
```
I'm not familiar with how S3 works, but given the example above, object name would be `"deploys/my-artifact-1.0.0.tgz"`. Is that correct?

If you are trying to get the filename here, perhaps you should try `File.basename` which would give you `"my-artifact-1.0.0.tgz"`

jonathanq replied:
S3 doesn't really have any notion of "files" and "folders".  Just objects.  While many tools (including the AWS Management Console) shows it as a folder-like structure, you fetch objects by their key which includes the full path (minus the bucket) to the object.

```
s3://bucket-name/some/path/to/file.tar.gz
```

So to get the `file.tar.gz` you need to request the object by its key: `some/path/to/file.tar.gz`

KAllan357 commented on libraries/chef_artifact.rb:
```diff
@@ -55,19 +57,28 @@ def windows?
       # 
       # @return [Chef::DataBagItem] the data bag item
-      def nexus_config_for(node)
+      def config_for(node, source)
```
I think this would be more clear if it were renamed to `data_bag_config_for`

jonathanq replied:
I'll rename it - good idea.

KAllan357 commented on libraries/chef_artifact.rb:
```diff
@@ -55,19 +57,28 @@ def windows?
+        return data_bag_item if DATA_BAG_NEXUS == source
+
+        # no config found for source
+        return {}
```
You can omit the return statement here.

KAllan357 commented:
Thanks for this pull, I think overall this looks pretty good. I think it would be useful if someone else took a look at this pull as well. @reset or @ivey want to weigh in?

jonathanq committed:
Fixed dependencies for aws-sdk (added gcc/make, changed action) and renamed config_for method

jonathanq commented:
Fixed up the changes that @KAllan357 suggested.

andrewGarson commented on libraries/chef_artifact.rb:
```diff
@@ -164,6 +222,16 @@ def from_nexus?(location)
+      #
+      # @return [Boolean] true when the location matches s3
+      def from_s3?(location)
+        location =~ URI::regexp('s3')
```
Could you match this a little more specifically?

I worry that somebody might try to pull down from a site like http://files3.something.com or some other false match.

jonathanq replied:
Sure - i'll add some spec tests around that as well to ensure it only catches the s3:// urls.

andrewGarson commented:
:+1:  aside from my comment about that regexp

jonathanq committed:
Refactored the from_<source>? methods to make them actually return a Boolean (as their docs stated) and added spec tests.  Also removed duplicated methods in the LWRPs

ivey commented on README.md:
```diff
@@ -128,7 +128,7 @@ create   | Download the artifact file    | Yes
 ---------              |-------------                                                                         |-----    |--------
 path                   | The path to download the artifact to                                                 | String  | name
-location               | The location to the artifact file. Either a nexus identifier or URL                  | String  |
+location               | The location to the artifact file. Either a nexus identifier, S3 path or URL        | String  |
```
:nitpick: spacing on the table is off by 1

jonathanq committed:
Fixed spacing on README.md and missed from_s3? from refactoring

jonathanq commented:
I cleaned up the `from_s3/from_nexus/from_http` calls.  Removed the duplicated calls from the LWRPs and just referenced the `Chef::Artifact` ones and added tests for them. 

The `URI::regexp` method being used already handles the case of checking the URI scheme only at the beginning - so won't match s3 in the domain name.  But added tests to be sure.  And I fixed up the spacing in the `README.md`

ivey commented on README.md:
```diff
@@ -178,7 +178,72 @@ Your data bag should look like the following:
+      ]
+    }
+
+If you wish to provide your AWS credentials in a data_bag, the format is:
```
double explanation comment here

ivey commented:
This looks solid to me

jonathanq committed:
Removed duplicated text

KAllan357 commented:
Thanks @jonathanq for your contribution! Just released this to the Opscode community site as version 1.7.0.
